lib/teleterm/daemon/gateway_cert_reissuer.go

```diff
@@ -0,0 +1,186 @@
+	err := r.reissueCert(ctx, gateway, dbCertReissuer)
+	if err != nil {
+		r.notifyAppAboutError(ctx, err, gateway)
+
+		// Return the error to the alpn.LocalProxy's middleware.
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (r *GatewayCertReissuer) reissueCert(ctx context.Context, gateway *gateway.Gateway, dbCertReissuer DBCertReissuer) error {
+	// Make the first attempt at reissuing the cert.
+	//
+	// It might happen that the db cert has expired but the user cert is still active, allowing us to
+	// obtain a new db cert without having to relogin first.
+	//
+	// This can happen if the user cert was refreshed by anything other than the gateway itself. For
+	// example, if you execute `tsh ssh` within Connect after your cert expires or there are two
+	// gateways that subsequently go through this flow.
+	err := r.reissueAndReloadGatewayCert(ctx, gateway, dbCertReissuer)
+
+	if err == nil {
+		return nil
+	}
+
+	// Do not ask for relogin if the error cannot be resolved with relogin.
+	if !client.IsErrorResolvableWithRelogin(err) {
+		return trace.Wrap(err)
+	}
+
+	clusterURI, err := uri.ParseClusterURI(gateway.TargetURI())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	rootClusterURI := clusterURI.GetRootClusterURI().String()
+
+	err = r.requestReloginFromElectronApp(ctx,
+		&api.ReloginRequest{
+			RootClusterUri: rootClusterURI,
+			Reason: &api.ReloginRequest_GatewayCertExpired{
+				GatewayCertExpired: &api.GatewayCertExpired{
+					GatewayUri: gateway.URI().String(),
+					TargetUri:  gateway.TargetURI(),
+				},
+			},
+		})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = r.reissueAndReloadGatewayCert(ctx, gateway, dbCertReissuer)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (r *GatewayCertReissuer) reissueAndReloadGatewayCert(ctx context.Context, gateway *gateway.Gateway, dbCertReissuer DBCertReissuer) error {
+	err := dbCertReissuer.ReissueDBCerts(ctx, gateway.RouteToDatabase())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = gateway.ReloadCert()
+	return trace.Wrap(err)
+}
+
+func (r *GatewayCertReissuer) requestReloginFromElectronApp(ctx context.Context, req *api.ReloginRequest) error {
+	// If you're changing the timeout, remember to update the error message for codes.DeadlineExceeded below.
+	timeoutCtx, cancelTshdEventsCtx := context.WithTimeout(ctx, time.Minute)
+	defer cancelTshdEventsCtx()
+
+	// The Electron app cannot display two login modals at the same time, so we have to cut short any
+	// concurrent relogin requests.
+	if !r.reloginMu.TryLock() {
+		return trace.AlreadyExists("another relogin request is in progress")
+	}
+	defer r.reloginMu.Unlock()
+
+	_, err := r.TSHDEventsClient.Relogin(timeoutCtx, req)
+
+	if err != nil {
+		if status.Code(err) == codes.DeadlineExceeded {
+			return trace.Wrap(err, "the user did not refresh the session within 1 minute")
+		}
```

Comment:
```suggestion
	const reloginUserTimeout = time.Minute
	timeoutCtx, cancelTshdEventsCtx := context.WithTimeout(ctx, reloginUserTimeout)
	defer cancelTshdEventsCtx()

	// The Electron app cannot display two login modals at the same time, so we have to cut short any
	// concurrent relogin requests.
	if !r.reloginMu.TryLock() {
		return trace.AlreadyExists("another relogin request is in progress")
	}
	defer r.reloginMu.Unlock()

	_, err := r.TSHDEventsClient.Relogin(timeoutCtx, req)

	if err != nil {
		if status.Code(err) == codes.DeadlineExceeded {
			return trace.Wrap(err, "the user did not refresh the session within %s", reloginUserTimeout.String())
		}
```

Reply:
Oh cool, I didn't know that there's a built-in way to format the duration and I didn't want to write custom code for that.